flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosFlinkResourceManager.java

```diff
@@ -0,0 +1,755 @@
+		// we do not unregister, but cause a hard fail of this process, to have it
+		// restarted by the dispatcher
+		LOG.error("FATAL ERROR IN MESOS APPLICATION MASTER: " + message, error);
+		LOG.error("Shutting down process");
+
+		// kill this process, this will make an external supervisor (the dispatcher) restart the process
+		System.exit(EXIT_CODE_FATAL_ERROR);
+	}
+
+	// ------------------------------------------------------------------------
+	//  Worker Management
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Recover framework/worker information persisted by a prior incarnation of the RM.
+	 */
+	private void recoverWorkers() throws Exception {
+		// if this application master starts as part of an ApplicationMaster/JobManager recovery,
+		// then some worker tasks are most likely still alive and we can re-obtain them
+		final List<MesosWorkerStore.Worker> tasksFromPreviousAttempts = workerStore.recoverWorkers();
+
+		if (!tasksFromPreviousAttempts.isEmpty()) {
+			LOG.info("Retrieved {} TaskManagers from previous attempt", tasksFromPreviousAttempts.size());
+
+			List<Tuple2<TaskRequest,String>> toAssign = new ArrayList<>(tasksFromPreviousAttempts.size());
+			List<LaunchableTask> toLaunch = new ArrayList<>(tasksFromPreviousAttempts.size());
+
+			for (final MesosWorkerStore.Worker worker : tasksFromPreviousAttempts) {
+				LaunchableMesosWorker launchable = createLaunchableMesosWorker(worker.taskID());
+
+				switch(worker.state()) {
+					case New:
+						workersInNew.put(extractResourceID(worker.taskID()), worker);
+						toLaunch.add(launchable);
+						break;
+					case Launched:
+						workersInLaunch.put(extractResourceID(worker.taskID()), worker);
+						toAssign.add(new Tuple2<>(launchable.taskRequest(), worker.hostname().get()));
+						break;
+					case Released:
+						workersBeingReturned.put(extractResourceID(worker.taskID()), worker);
+						break;
+				}
+				taskRouter.tell(new TaskMonitor.TaskGoalStateUpdated(extractGoalState(worker)), self());
+			}
+
+			// tell the launch coordinator about prior assignments
+			if(toAssign.size() >= 1) {
+				launchCoordinator.tell(new LaunchCoordinator.Assign(toAssign), self());
+			}
+			// tell the launch coordinator to launch any new tasks
+			if(toLaunch.size() >= 1) {
+				launchCoordinator.tell(new LaunchCoordinator.Launch(toLaunch), self());
+			}
+		}
+	}
+
+	/**
+	 * Plan for some additional workers to be launched.
+	 *
+	 * @param numWorkers The number of workers to allocate.
+     */
+	@Override
+	protected void requestNewWorkers(int numWorkers) {
+
+		try {
+			List<TaskMonitor.TaskGoalStateUpdated> toMonitor = new ArrayList<>(numWorkers);
+			List<LaunchableTask> toLaunch = new ArrayList<>(numWorkers);
+
+			// generate new workers into persistent state and launch associated actors
+			for (int i = 0; i < numWorkers; i++) {
+				MesosWorkerStore.Worker worker = MesosWorkerStore.Worker.newTask(workerStore.newTaskID());
```

Comment:
Worker is the logical term used by the base class; task is the concrete term used by Mesos.   I strove for some internal consistency when dealing with logical vs concrete.